src/coreComponents/meshUtilities/CPMesh/CPMeshBuilder.cpp

```diff
@@ -0,0 +1,401 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 Total, S.A
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file CPMeshBuilder.cpp
+ */
+
+#include "CPMeshBuilder.hpp"
+
+#include "codingUtilities/Utilities.hpp"
+#include "mpiCommunications/MpiWrapper.hpp"
+#include "rajaInterface/GEOS_RAJA_Interface.hpp"
+#include "meshUtilities/CPMesh/utilities/GeometryUtilities.hpp"
+#include "meshUtilities/CPMesh/utilities/OutputUtilities.hpp"
+
+namespace geosx
+{
+
+namespace CPMesh
+{
+
+using namespace GeometryUtilities;
+using namespace OutputUtilities;
+
+CPMeshBuilder::CPMeshBuilder( string const & name )
+  :
+  m_meshParser( name ),
+  m_meshComm( name ),
+  m_meshName( name )
+{}
+
+void CPMeshBuilder::buildMesh( Path const & filePath )
+{
+  localIndex const iRank = MpiWrapper::commRank( MPI_COMM_GEOSX );
+
+  if( iRank == 0 )
+  {
+    localIndex nX = 0;
+    localIndex nY = 0;
+    localIndex nZ = 0;
+
+    // read SPECGRID/DIMENS to get the number of cells in each direction
+    m_meshParser.readNumberOfCells( filePath, nX, nY, nZ );
+    m_meshDims.defineDomainDimensions( nX, nY, nZ );
+
+    // TODO: read ACTNUM for the inertial partitioning
+  }
+
+  // rank 0 partitions the mesh and sends partitions to other ranks
+  // rank 0 sets up the communication pattern
+  m_meshComm.setupMPIPartition( m_meshDims );
+
+  // each rank reads its part of the mesh
+  m_meshParser.readMesh( filePath, m_meshDims );
+
+  // post-process the mesh to make it conforming
+  postProcessMesh();
+}
+
+void CPMeshBuilder::postProcessMesh()
+{
+  // for each active cell, compute the position of the eight CPG vertices
+  // at this point, CPG vertices have not been filtered yet
+  buildCornerPointCells();
```

Comment:
CPG means Corner Point Grid. In the Corner Point format, each cell has 8 corner point vertices (the `cpVertices`) of the code, so that is what these `CPG vertices` in the comment refer to.